internal/clients/telegram/client.go

```diff
@@ -0,0 +1,282 @@
+package telegram
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"sync"
+	"time"
+)
+
+// A client to interact with the telegram API. Start the client and then call Stop to stop it gracefully.
+// The client may take some time to shutdown if it has work to do.
+//
+// # Example
+//
+//	client := telegram.NewClient("api.telegram.org", "TOKEN")
+//	client.Start(10) // 10 threads
+//
+//	if shouldStop { // whatever the trigger for stop is
+//	    client.Stop()
+//	}
+type Client struct {
+	host           string // URL of the API server, without `https://`. E.g. `api.telegram.org`
+	basePath       string // `basePath + endpointPath` for making requests. Constructed from `"bot"+token`
+	client         http.Client
+	wg             sync.WaitGroup     //Used to make sure all processor threads are done
+	stopProcessing context.CancelFunc // Triggers the shutdown
+}
+
+// Creates a new client.
+//
+// `host` is the address to the server, without `https://`
+//
+// `token` is the bot token for the API
+func NewClient(host, token string) Client {
+	return Client{
+		host:     host,
+		basePath: "bot" + token,
+		client:   http.Client{},
+	}
+}
+
+/*
+Starts the client in the background. It will fetch updates via `Client.getUpdates()`
+and sends them to a queue. Queue items are recieved by one of `Client.processUpdates()`
+which are going to process all updates in parallel
+
+`goroutines` specifies how many threads to use in parallel.
+
+# Panics
+
+This function considers 0 or less goroutines as fatal.
+*/
+func (c *Client) Start(goroutines uint) {
+	if goroutines < 1 {
+		panic(fmt.Sprintf(
```

Comment:
do you really need a panic here?

Reply:
Are you implying that 0 should just mean 1?

Reply:
My question is in when this happened, should the service immediately crash?

Reply:
I assume this only runs once on startup so it has 2 options:
- Crash the server and tell the programmer they made an error
- Silently assume what to do with the 0 or log this error.

I think it should be a fatal error to pass 0.

Reply:
If you have another way of dealing with invalid values then please tell me.

One thing that I don't think is a good idea is returning error from Start. This is because when you get the return value you'd usually think its the result of the function, not argument value errors. For these cases panics are better because they dont return a value. And there's a doc comment explaining when the funciton panics and what the argument means so I dont see any issue here.

Reply:
I didn't see a problem in returning error. Returning an error from Start can be confusing because it may be mistaken as the result of the function is not a strong justification for using panics. In this case, is considered a standard and idiomatic way to handle invalid values. By returning an error, the caller of the function can handle the error gracefully.


also, I recommended you use log.Fatal